pkg/controller/admissionchecks/multikueue/multikueuecluster.go

```diff
@@ -0,0 +1,345 @@
+		return reconcile.Result{}, c.updateStatus(ctx, cluster, false, "BadConfig", err.Error())
+	}
+
+	if err := c.setRemoteClientConfig(ctx, cluster.Name, kubeConfig); err != nil {
+		log.Error(err, "setting kubeconfig")
+		return reconcile.Result{}, c.updateStatus(ctx, cluster, false, "ClientConnectionFailed", err.Error())
+	}
+	return reconcile.Result{}, c.updateStatus(ctx, cluster, true, "Active", "Connected")
+}
+
+func (c *clustersReconciler) getKubeConfig(ctx context.Context, ref *kueuealpha.KubeConfig) ([]byte, bool, error) {
+	sec := corev1.Secret{}
+	secretObjKey := types.NamespacedName{
+		Namespace: c.configNamespace,
+		Name:      ref.Location,
+	}
+	err := c.client.Get(ctx, secretObjKey, &sec)
+	if err != nil {
+		return nil, !apierrors.IsNotFound(err), err
+	}
+
+	kconfigBytes, found := sec.Data[kueuealpha.MultiKueueConfigSecretKey]
+	if !found {
+		return nil, false, fmt.Errorf("key %q not found in secret %q", kueuealpha.MultiKueueConfigSecretKey, ref.Location)
+	}
+
+	return kconfigBytes, false, nil
+}
+
+func (c *clustersReconciler) updateStatus(ctx context.Context, cluster *kueuealpha.MultiKueueCluster, active bool, reason, message string) error {
+	newCondition := metav1.Condition{
+		Type:    kueuealpha.MultiKueueClusterActive,
+		Status:  metav1.ConditionFalse,
+		Reason:  reason,
+		Message: message,
+	}
+	if active {
+		newCondition.Status = metav1.ConditionTrue
+	}
+
+	// if the cluster is connected and remains that way, skip the status update
+	oldCondition := apimeta.FindStatusCondition(cluster.Status.Conditions, kueuealpha.MultiKueueClusterActive)
+	if active && oldCondition != nil && oldCondition.Status == metav1.ConditionTrue {
+		return nil
+	}
+
+	apimeta.SetStatusCondition(&cluster.Status.Conditions, newCondition)
+	return c.client.Status().Update(ctx, cluster)
+}
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=multikueueclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=multikueueclusters/status,verbs=get;update;patch
+
+func newClustersReconciler(c client.Client, namespace string) *clustersReconciler {
+	return &clustersReconciler{
+		client:          c,
+		configNamespace: namespace,
+		clients:         make(map[string]*remoteClient),
+		wlUpdateCh:      make(chan event.GenericEvent, 10),
+	}
+}
+
+func (c *clustersReconciler) setupWithManager(mgr ctrl.Manager) error {
+	err := mgr.Add(c)
+	if err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kueuealpha.MultiKueueCluster{}).
+		Watches(&corev1.Secret{}, &secretHandler{client: c.client}).
+		Complete(c)
+}
+
+type secretHandler struct {
+	client client.Client
+}
+
+var _ handler.EventHandler = (*secretHandler)(nil)
+
+func (s *secretHandler) Create(ctx context.Context, event event.CreateEvent, q workqueue.RateLimitingInterface) {
+	if err := s.queue(ctx, event.Object, q); err != nil {
```

Comment:
Can we check `event.Object` here? Or can we output error logs in `queue` function?

Reply:
@trasc Yes, that's right, but I still think that the additional key/value is odd in that case.

> "secret", klog.KObj(event.ObjectOld)

So, I still would suggest applying my comment.

Reply:
@tenzen-y done